File: cronos-service/src/main/java/br/jus/tjrs/cronos/app/model/service/CategoriaItemService.java
package br.jus.tjrs.cronos.app.model.service;

import java.util.List;

import javax.ejb.Local;

import br.jus.tjrs.cronos.app.model.domain.TipoArvore;
import br.jus.tjrs.cronos.app.model.dto.AutoCompleteDTO;
import br.jus.tjrs.cronos.app.model.dto.CarregarCategoriaDTO;
import br.jus.tjrs.cronos.app.model.dto.CategoriaItemDTO;
import br.jus.tjrs.cronos.app.model.dto.CloneCategoriaItemDTO;
import br.jus.tjrs.cronos.app.model.dto.IconeDTO;
import br.jus.tjrs.cronos.app.model.dto.PesquisaGeralDTO;
import br.jus.tjrs.cronos.app.model.dto.RetornoPesquisaGeralDTO;
import br.jus.tjrs.cronos.app.model.dto.SaveNodeDTO;
import br.jus.tjrs.cronos.app.model.entity.CategoriaItemEntity;
import br.jus.tjrs.cronos.app.model.entity.GrupoTrabalhoEntity;
import br.jus.tjrs.cronos.core.CronosException;
import br.jus.tjrs.cronos.core.model.service.EntityService;

@Local
public interface CategoriaItemService extends EntityService<Long, CategoriaItemEntity>
{
   CategoriaItemDTO carregarCategoria(CarregarCategoriaDTO carregarCategoria) throws CronosException;

   List<CategoriaItemEntity> findAllByPai(Long idCategoria) throws CronosException;

   List<CategoriaItemEntity> findAllByGrupoTrabalho(Long idGrupoTrabalho) throws CronosException;

   CategoriaItemEntity salvarTopico(CategoriaItemDTO categoriaItemDTO) throws CronosException;

   List<CategoriaItemEntity> listByPartDescription(AutoCompleteDTO autocomplete) throws CronosException;

   CategoriaItemEntity alterarIcone(IconeDTO icone) throws CronosException;

   List<RetornoPesquisaGeralDTO> buscarPesquisaGeral(PesquisaGeralDTO part) throws CronosException;

   CategoriaItemEntity cloneItem(CloneCategoriaItemDTO categoria) throws CronosException;

   List<CategoriaItemEntity> listNodeByPartDescription(AutoCompleteDTO autocomplete) throws CronosException;

   CategoriaItemEntity saveNode(SaveNodeDTO node) throws CronosException;

   List<CategoriaItemEntity> findAllByGrupoTrabalhoTipoArvore(Long id, TipoArvore tipoArvore) throws CronosException;

   void createFirstTreeView(GrupoTrabalhoEntity grupoTrabalhoNovo, GrupoTrabalhoEntity grupoTrabalhoAntigo)
            throws CronosException;

   List<CategoriaItemEntity> listTopicosByIdModelo(Long id) throws CronosException;
}

File: cronos-frontend/src/scripts/app/services/entity/categoriaItemService.js
angular.module('cronos')
    .factory('ciService', ['EntityService', function CategoriaItemServiceFactory(EntityService) {
    'use strict';

    var CategoriaItemService = $class(EntityService, {

        constructor: function() {
            EntityService.call(this, {
                type: 'categoriaItem'
            });
        },

        /**
         * @param
         * {
         * 		tipoArvore = ('T'/'M'/'S'),
         * 		idCategoria,
         * 		idGrupoTrabalho
         * }
         *
         * @returns
         * {
         *  	categoria = <CategoriaItemEntity> {
         * 			id;
         * 			titulo;
         * 			icone,
         * 			cor,
         * 			idIcone,
         * 			ordem,
         * 			idPai,
         * 			idGrupoTrabalho,
         * 			descricaoArvore,
         * 			compartilhar = ('S'/'N'),
         * 			tipo, ('N'/'E'/'I')
         * 			arvore, ('T'/'M'/'S')
         * 			iconePersonal = <IconesPersonalEntity> {
         * 				id,
         * 				imagem,
         * 				tipo
         * 			},
         * 			tipoCategoria = <TipoCategoriaEntity> {
         * 				id,
         * 				nome
         * 			},
         * 			itens = [<CategoriaItemEntity>],
         * 			taxonomiasEtiqueta = [<TaxonomiaEtiquetaGTEntity>],
         * 			elementos = [<ElementoIGTEntity>]
         * 		},
         *		categoriasModelo = [<CategoriaItemEntity>]
         * }
         * */
        carregarCategoria: function(carregarCategoria) {
            return this._post('/carregarCategoria', carregarCategoria);
        },

        /**
         * @param
         * {
         * 		part,
         * 		idGrupoTrabalho
         * }
         *
         * @returns
         * {
         * 		[<CategoriaItemEntity>]
         * }
         * */
        listByPartDescription : function(autoCompleteDTO) {
            return this._post('/listByPartDescription', autoCompleteDTO);
        },

        /**
         * @param
         * {
         * 		part,
         * 		idGrupoTrabalho
         * }
         *
         * @returns
         * {
         * 		[<CategoriaItemEntity>]
         * }
         * */
        listNodeByPartDescription : function(autoCompleteDTO) {
            return this._post('/listNodeByPartDescription', autoCompleteDTO);
        },

        /**
         * @param
         * {
         * 		titulo,
         * 		idGrupoTrabalho,
         *      usuarioAtualizacao,
         *      tipoArvore, ('T'/'M'/'S')
         * }
         *
         * @returns
         * {
         * 		<CategoriaItemEntity>
         * }
         * */
        saveNode : function(node) {
            return this._post('/saveNode', node);
        },

        /**
         * @param
         * {
         * 		part,
         * 		tipoArvore = ('T'/'M'/'S'),
         * 		idGrupoTrabalho,
         * }
         *
         * @returns
         * [{
         * 		id,
         * 		titulo,
         *  	arvore = ('T'/'M'/'S'),
         *  	label = [{String}],
         *		excluir = (true/false),
         *		editar = (true/false),
         *		revisar = (true/false),
         *		adicionar = (true/false),
         *		visualizar = (true/false),
         * }]
         * */
        buscarPesquisaGeral : function(pesquisaGeralDTO) {
            return this._post('/buscarPesquisaGeral', pesquisaGeralDTO);
        },

        /**
         * @param
         * {
         *  	categoria = <CategoriaItemEntity>,
         *  	taxonomiasEtiqueta = [<TaxonomiaEtiquetaGTEntity>],
         *  	taxonomiasCategoria = [<TaxonomiaCategoriaEntity>],
         *  	taxonomiasCategoriaNovas = [<TaxonomiaCategoriaEntity>,
         *  	elementos = [<ElementoIGTEntity>]
         * }
         *
         * @returns
         * <CategoriaItemEntity> {
         * 		id;
         * 		titulo;
         * 		icone,
         * 		cor,
         * 		idIcone,
         * 		ordem,
         * 		idPai,
         * 		idGrupoTrabalho,
         * 		descricaoArvore,
         * 		compartilhar = ('S'/'N'),
         * 		TipoNodo tipo, ('N'/'E'/'I')
         * 		TipoArvore arvore, ('T'/'M'/'S')
         * 		iconePersonal = <IconesPersonalEntity> {
         * 			id,
         * 			imagem,
         * 			tipo
         * 		},
         * 		tipoCategoria = <TipoCategoriaEntity> {
         * 			id,
         * 			nome
         * 		},
         * 		itens = [<CategoriaItemEntity>],
         * 		taxonomiasEtiqueta = [<TaxonomiaEtiquetaGTEntity>],
         * 		elementos = [<ElementoIGTEntity>]
         * }
         * */
        salvarTopico: function(topico) {
            return this._post('/salvarTopico', topico);
        },

        /**
         * @param
         * {
         * 		id,
         * 		icone,
         * 		cor,
         * 		imagem = {
         * 			filesize,
         * 			filetype,
         * 			filename,
         * 			base64
         * 		}
         * }
         *
         * @returns
         * <CategoriaItemEntity> {
         * 		id;
         * 		titulo;
         * 		icone,
         * 		cor,
         * 		idIcone,
         * 		ordem,
         * 		idPai,
         * 		idGrupoTrabalho,
         * 		descricaoArvore,
         * 		compartilhar = ('S'/'N'),
         * 		TipoNodo tipo, ('N'/'E'/'I')
         * 		TipoArvore arvore, ('T'/'M'/'S')
         * 		iconePersonal = <IconesPersonalEntity>{
         * 			id,
         * 			imagem,
         * 			tipo
         * 		},
         * 		tipoCategoria = <TipoCategoriaEntity>{
         * 			id,
         * 			nome
         * 		},
         * 		itens = [<CategoriaItemEntity>],
         * 		taxonomiasEtiqueta = [<TaxonomiaEtiquetaGTEntity>],
         * 		elementos = [<ElementoIGTEntity>]
         * }
         * */
        alterarIcone: function(icone) {
            return this._post('/alterarIcone', icone);
        },

        cloneItem: function(item) {
            return this._post('/cloneItem', item);
        },

        listTopicosByIdModelo: function(id) {
            return this._get('/listTopicosByIdModelo/' + id);
        }
    });

    return new CategoriaItemService();
}]);
